modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/certificatevalidation/CertificateVerificationManager.java

```diff
@@ -0,0 +1,282 @@
+                             InvalidKeyException | NoSuchProviderException e) {
+                        // Unable to verify the signature. Check with the next certificate in the next loop traversal.
+                    }
+                }
+            } else {
+                X509Certificate cachedIssuerCert = certCache.getCacheValue(peerCert.getSerialNumber().toString());
+                try {
+                    peerCert.verify(cachedIssuerCert.getPublicKey());
+                } catch (SignatureException | CertificateException | NoSuchAlgorithmException |
+                         InvalidKeyException |
+                         NoSuchProviderException e) {
+                    // Unable to verify the signature.
+                    throw new CertificateVerificationException("Unable to verify the signature of the certificate.");
+                }
+            }
+        }
+
+        long start = System.currentTimeMillis();
+
+        OCSPCache ocspCache = OCSPCache.getCache();
+        ocspCache.init(cacheSize, cacheDelayMins);
+        CRLCache crlCache = CRLCache.getCache();
+        crlCache.init(cacheSize, cacheDelayMins);
+
+        RevocationVerifier[] verifiers = {new OCSPVerifier(ocspCache), new CRLVerifier(crlCache)};
+
+        for (RevocationVerifier verifier : verifiers) {
+            try {
+                if (isFullCertChainValidationEnabled) {
+
+                    if (isCertExpiryValidationEnabled) {
+                        log.debug("Validating certificate chain for expiry");
+                        if (isExpired(convertedCertificates)) {
+                            throw new CertificateVerificationException("One of the provided certificates are expired");
+                        }
+                    }
+
+                    log.debug("Doing full certificate chain validation");
+                    CertificatePathValidator pathValidator = new CertificatePathValidator(convertedCertificates,
+                            verifier);
+                    pathValidator.validatePath();
+                    log.info("Path verification Successful. Took " + (System.currentTimeMillis() - start) + " ms.");
+                } else {
+
+                    if (isCertExpiryValidationEnabled) {
+                        log.debug("Validating the client certificate for expiry");
+                        if (isExpired(convertedCertificates)) {
+                            throw new CertificateVerificationException("The provided certificate is expired");
+                        }
+                    }
+
+                    log.debug("Validating client certificate with the issuer certificate retrieved from" +
+                            "the trust store");
+                    verifier.checkRevocationStatus(peerCert, issuerCert);
+                }
+                return;
+            } catch (Exception e) {
+                log.info(verifier.getClass().getSimpleName() + " failed.");
```

Comment:
shouldn't we make this an error log?

Reply:
I think this is not written as an error log since if this catch block is called, the next verifier will be hit. If we print an error log, it will give the user the wrong idea. If both verifiers fail, an exception like below is thrown at the end of the method.

```
throw new CertificateVerificationException("Path Verification Failed for both OCSP and CRL");
```

Reply:
then we shouldnt print a info log too.

Reply:
Fixed with https://github.com/wso2/wso2-synapse/pull/2119/commits/d02fc9094da18bb166796929d8009da570dfb0b2